core/modules/wbase/SendChannelShared.cc

```diff
@@ -0,0 +1,223 @@
+
+namespace lsst {
+namespace qserv {
+namespace wbase {
+
+SendChannelShared::Ptr SendChannelShared::create(SendChannel::Ptr const& sendChannel,
+                                                 wcontrol::TransmitMgr::Ptr const& transmitMgr)  {
+    auto scs = shared_ptr<SendChannelShared>(new SendChannelShared(sendChannel, transmitMgr));
+    return scs;
+}
+
+
+void SendChannelShared::setTaskCount(int taskCount) {
+    _taskCount = taskCount;
+}
+
+
+bool SendChannelShared::transmitTaskLast(StreamGuard sLock, bool inLast) {
+    /// _caller must have locked _streamMutex before calling this.
+    if (not inLast) return false; // This wasn't the last message buffer for this task, so it doesn't matter.
+    ++_lastCount;
+    bool lastTaskDone = _lastCount >= _taskCount;
+    return lastTaskDone;
+}
+
+
+bool SendChannelShared::kill(StreamGuard sLock) {
+    LOGS(_log, LOG_LVL_DEBUG, "SendChannelShared::kill() called");
+    bool ret = _sendChannel->kill();
+    _lastRecvd = true;
+    return ret;
+}
+
+string SendChannelShared::makeIdStr(int qId, int jId) {
+    string str("QID" + (qId == 0 ? "" : to_string(qId) + "#" + to_string(jId)));
+    return str;
+}
+
+
+bool SendChannelShared::addTransmit(bool cancelled, bool erred, bool last, bool largeResult,
+                                    TransmitData::Ptr const& tData, int qId, int jId) {
+    QSERV_LOGCONTEXT_QUERY_JOB(qId, jId);
+    assert(tData != nullptr);
+
+    // This lock may be held for a very long time.
+    std::unique_lock<std::mutex> qLock(_queueMtx);
+    _transmitQueue.push(tData);
+
+    // If _lastRecvd is true, the last message has already been transmitted and
+    // this SendChannel is effectively dead.
+    bool reallyLast = _lastRecvd;
+    string idStr(makeIdStr(qId, jId));
+
+    // If something bad already happened, just give up.
+    if (reallyLast || isDead()) {
+        // If there's been some kind of error, make sure that nothing hangs waiting
+        // for this.
+        LOGS(_log, LOG_LVL_WARN, "addTransmit getting messages after isDead or reallyLast " << idStr);
+        _lastRecvd = true;
+        return false;
+    }
+    {
+        lock_guard<mutex> streamLock(streamMutex);
+        reallyLast = transmitTaskLast(streamLock, last);
+    }
+
+    if (reallyLast || erred || cancelled) {
+        _lastRecvd = true;
+    }
+
+    // If this is reallyLast or at least 2 items are in the queue, the transmit can happen
+    if (reallyLast || _transmitQueue.size() >= 2) {
+        bool scanInteractive = tData->scanInteractive;
+        // If there was an error, give this high priority.
+        if (erred) scanInteractive = true;
+        int czarId = tData->czarId;
+        // Don't wait if cancelled or error.
+        if (erred || cancelled) {
+            return _transmit(erred, largeResult, czarId);
+        }
+        // Limit the number of concurrent transmits.
+        wcontrol::TransmitLock transmitLock(*_transmitMgr, scanInteractive, largeResult, czarId);
```

Comment:
I would guess this is another place where a wait will be imposed, yes?

Reply:
This is the only place. There are lots of different SendChannelShared objects all fighting to send their results back to the czars. TranmitLock/TransmitMgr try to keep the amount of traffic from this worker at reasonable levels. Basically 50 per czar, with each czar having its own count.